ambry-tools/src/main/java/com.github.ambry/tools/admin/ConcurrencyTestTool.java

```diff
@@ -474,6 +505,81 @@ public void onCompletion(Void result, Exception exception) {
   }
 
   /**
+   * Class to assist in collecting and reporting metrics
+   */
+  class MetricsCollector {
+    private String operation;
+    private final long measurementIntervalInSecs;
+    private ArrayList<Long> requestLatencies;
+    private AtomicLong timePassedInMs = new AtomicLong(0);
+    private AtomicLong numberOfOperations = new AtomicLong(0);
+    private AtomicLong maxLatencyInMs = new AtomicLong(0);
+    private AtomicLong minLatencyInMs = new AtomicLong(Long.MAX_VALUE);
+    private AtomicLong totalLatencyInMs = new AtomicLong(0);
+
+    MetricsCollector(long measurementIntervalInSecs, String operation) {
+      this.measurementIntervalInSecs = measurementIntervalInSecs;
+      requestLatencies = new ArrayList<Long>();
+      this.operation = operation;
+    }
+
+    /**
+     * Updates the wall clock time spent so far. This is used for reporting purpose.
+     * For every {@code measurementIntervalInSecs}, we report the metrics and clear the stats.
+     * @param timePassedInMs the time spent so far in Millisecs
+     */
+    synchronized void updateTimePassedSoFar(long timePassedInMs) {
+      this.timePassedInMs.addAndGet(timePassedInMs);
+      if (this.timePassedInMs.get() >= measurementIntervalInSecs * SystemTime.MsPerSec) {
+        reportMetrics();
+        numberOfOperations.set(0);
+        this.timePassedInMs.set(0);
+        requestLatencies.clear();
+        maxLatencyInMs.set(0);
+        minLatencyInMs.set(Long.MAX_VALUE);
+        totalLatencyInMs.set(0);
+      }
+    }
+
+    /**
+     * Reports the stats for latencies like avg, 95th, 99th, min and max
+     */
+    void reportMetrics() {
+      if (requestLatencies.size() > 5) {
+        Collections.sort(requestLatencies);
+        int index99 = (int) (requestLatencies.size() * 0.99) - 1;
+        int index95 = (int) (requestLatencies.size() * 0.95) - 1;
+        String message =
+            operation + ":totalOps=" + numberOfOperations + ", 99thInMs=" + (double) requestLatencies.get(index99)
+                + ", 95thInMs=" + (double) requestLatencies.get(index95) + ", AvgInMs=" + (
+                ((double) totalLatencyInMs.get()) / numberOfOperations.get()) +
+                ", minInMs=" + minLatencyInMs.get() + ", maxInMs=" + (maxLatencyInMs.get());
+        logger.info(
```

Comment:
Should these line breaks be integrated into the beginning and end of `message` so that they are guaranteed to immediately precede/follow the log message?